session/sessionwants.go

```diff
@@ -0,0 +1,190 @@
+// limit). Returns the newly live wants.
+func (sw *sessionWants) GetNextWants(limit int, newWants []cid.Cid) []cid.Cid {
+	now := time.Now()
+
+	sw.Lock()
+	defer sw.Unlock()
+
+	// Add new wants to the fetch queue
+	for _, k := range newWants {
+		sw.toFetch.Push(k)
+	}
+
+	// Move CIDs from fetch queue to the live wants queue (up to the limit)
+	currentLiveCount := len(sw.liveWants)
+	toAdd := limit - currentLiveCount
+
+	var live []cid.Cid
+	for ; toAdd > 0 && sw.toFetch.Len() > 0; toAdd-- {
+		c := sw.toFetch.Pop()
+		live = append(live, c)
+		sw.liveWants[c] = now
+	}
+
+	return live
+}
+
+// PrepareBroadcast saves the current time for each live want and returns the
+// live want CIDs.
+func (sw *sessionWants) PrepareBroadcast() []cid.Cid {
+	now := time.Now()
+
+	sw.Lock()
+	defer sw.Unlock()
+
+	live := make([]cid.Cid, 0, len(sw.liveWants))
+	for c := range sw.liveWants {
+		live = append(live, c)
+		sw.liveWants[c] = now
+	}
+	return live
+}
+
+// CancelPending removes the given CIDs from the fetch queue.
+func (sw *sessionWants) CancelPending(keys []cid.Cid) {
+	sw.Lock()
+	defer sw.Unlock()
+
+	for _, k := range keys {
+		sw.toFetch.Remove(k)
+	}
+}
+
+// ForEachUniqDup iterates over each of the given CIDs and calls isUniqFn
+// if the session is expecting a block for the CID, or isDupFn if the session
+// has already received the block.
+func (sw *sessionWants) ForEachUniqDup(ks []cid.Cid, isUniqFn, isDupFn func()) {
+	sw.RLock()
+
+	for _, k := range ks {
+		if sw.unlockedIsWanted(k) {
+			isUniqFn()
+		} else if sw.pastWants.Has(k) {
+			isDupFn()
+		}
+	}
+
+	sw.RUnlock()
+}
+
+// LiveWants returns a list of live wants
+func (sw *sessionWants) LiveWants() []cid.Cid {
+	sw.RLock()
+	defer sw.RUnlock()
+
+	live := make([]cid.Cid, 0, len(sw.liveWants))
+	for c := range sw.liveWants {
+		live = append(live, c)
+	}
+	return live
+}
+
+// RandomLiveWant returns a randomly selected live want
+func (sw *sessionWants) RandomLiveWant() cid.Cid {
+	sw.RLock()
+	defer sw.RUnlock()
+
+	if len(sw.liveWants) == 0 {
+		return cid.Cid{}
+	}
+	i := rand.Intn(len(sw.liveWants))
+	// picking a random live want
+	for k := range sw.liveWants {
+		if i == 0 {
+			return k
+		}
+		i--
+	}
+	return cid.Cid{}
+}
+
+// Has live wants indicates if there are any live wants
+func (sw *sessionWants) HasLiveWants() bool {
+	sw.RLock()
+	defer sw.RUnlock()
+
+	return len(sw.liveWants) > 0
+}
+
+// IsWanted indicates if the session is expecting to receive the block with the
+// given CID
+func (sw *sessionWants) IsWanted(c cid.Cid) bool {
+	sw.RLock()
+	defer sw.RUnlock()
+
+	return sw.unlockedIsWanted(c)
+}
+
+// FilterInteresting filters the list so that it only contains keys for
+// blocks that the session is waiting to receive or has received in the past
+func (sw *sessionWants) FilterInteresting(ks []cid.Cid) []cid.Cid {
+	sw.RLock()
+	defer sw.RUnlock()
+
+	interested := make([]cid.Cid, 0, len(ks))
```

Comment:
I'd not pre-allocate here. It's unlikely that more than one session will be interested in the blocks.